ctest-next/templates/test.rs

```diff
@@ -82,7 +83,430 @@ mod generated_tests {
+    }
+
+    /// Tests whether the type alias `x` when passed to C and back to Rust remains unchanged.
+    ///
+    /// It checks if the size is the same as well as if the padding bytes are all in the
+    /// correct place.
+    pub fn roundtrip_{{ ident }}() {
+        use std::ffi::c_int;
+
+        type U = {{ ident }};
+        extern "C" {
+            fn __test_roundtrip_{{ ident }}(
+                size: i32, x: U, e: *mut c_int, pad: *const u8
+            ) -> U;
+        }
+        let pad = roundtrip_padding_{{ ident }}();
+        unsafe {
+            use std::mem::{MaybeUninit, size_of};
+            let mut error: c_int = 0;
+            let mut y = MaybeUninit::<U>::uninit();
+            let mut x = MaybeUninit::<U>::uninit();
+            let x_ptr = x.as_mut_ptr().cast::<u8>();
+            let y_ptr = y.as_mut_ptr().cast::<u8>();
+            let sz = size_of::<U>();
+            for i in 0..sz {
+                let c: u8 = (i % 256) as u8;
+                let c = if c == 0 { 42 } else { c };
+                let d: u8 = 255_u8 - (i % 256) as u8;
+                let d = if d == 0 { 42 } else { d };
+                x_ptr.add(i).write_volatile(c);
+                y_ptr.add(i).write_volatile(d);
+            }
+            let r: U = __test_roundtrip_{{ ident }}(sz as i32, x.assume_init(), &mut error, pad.as_ptr());
+            if error == 1 {
+                FAILED.store(true, Ordering::Relaxed);
+                return;
+            }
+            for (i, elem) in pad.iter().enumerate().take(size_of::<U>()) {
+                if *elem == 1 { continue; }
+                let rust = (*y_ptr.add(i)) as usize;
+                let c = (&r as *const _ as *const u8)
+                            .add(i).read_volatile() as usize;
+                if rust != c {
+                    eprintln!(
+                        "rust [{i}] = {rust} != {c} (C): C \"{{ ident }}\" -> Rust",
+                    );
+                    FAILED.store(true, Ordering::Relaxed);
+                }
+            }
+        }
+    }
+    {%- endif +%}
+    {%- endfor +%}
+
+    {%- for structure in ffi_items.structs() +%}
+    {%- let ident = structure.ident() +%}
+    {%- let fields = structure.public_fields() +%}
+
+    /// Test that the size and alignment of the aliased type is the same in both Rust and C.
+    /// 
+    /// This can fail if a different type is used on one side, and uses the built in size and
+    /// alignment functions to check.
+    pub fn size_align_{{ ident }}() {
+        extern "C" {
+            fn __test_size_{{ ident }}() -> u64;
+            fn __test_align_{{ ident }}() -> u64;
+        }
+        unsafe {
+            check_same(mem::size_of::<{{ ident }}>() as u64,
+                __test_size_{{ ident }}(), "{{ ident }} size");
+            check_same(mem::align_of::<{{ ident }}>() as u64,
+                __test_align_{{ ident }}(), "{{ ident }} align");
+        }
+    }
+
+    /// No idea what this does.
```

Comment:
Lol

The first bit is pretty easy: it just checks that the offsets of a field within a struct are the same, and that the field sizes are the same.

The second part makes an instance of the type and returns a pointer to the field, then checks it's the same in Rust. I'm not really sure what this is intended to test since it should be the same as `offset_of!`, but I guess it doesn't hurt to have.